### PR TITLE
Tests: Cluster consensus - Several duties

### PR DESCRIPTION
**Pull request by MatheusFranco99**
This PR adds tests for the `Committee` structure in the context of multiple duties execution.

## Tests
- [x] Sequenced decided duties
- [x] Sequenced happy flow duties
- [x] Shuffled decided duties
- [x] Shuffled happy flow duties with same validators
- [x] Shuffled happy flow duties with different validators
- [x] Failed duty than successful duty

## To-Do
- [x] Add test cases for 500 validators. Problem: `tests.json` of `ssv` package is getting bigger than 100MB (can't be added to GitHub)

### CODE REVIEW

File: ssv/spectest/tests/committee/test.go [hunk 1]
```diff
@@ -53,8 +53,8 @@ func (test *CommitteeSpecTest) RunAsPartOfMultiTest(t *testing.T) {
 		broadcastedRoots = append(broadcastedRoots, beaconNetwork.BroadcastedRoots...)
 	}
 
-	// test output message
-	testingutils.ComparePartialSignatureOutputMessages(t, test.OutputMessages, broadcastedMsgs, test.Committee.Operator.Committee)
+	// test output message (in asynchronous order)
+	testingutils.ComparePartialSignatureOutputMessagesInAsynchronousOrder(t, test.OutputMessages, broadcastedMsgs, test.Committee.Operator.Committee)
```
Comment (GalRogozinski) [inline ssv/spectest/tests/committee/test.go:57]: this is because of the shuffle tests?

**Review by GalRogozinski — COMMENTED** [review]
I don't know why, some of those tests remind me of already existing tests...

On the sync let's talk about the large JSON.
The options are to split it, use git LFS, or your idea that I need to understand better

File: ssv/spectest/tests/committee/multipleduty/sequenced_decided_duties.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+// SequencedDecidedDuties decides a sequence of duties
+func SequencedDecidedDuties() tests.SpecTest {
+
+	multiSpecTest := &committee.MultiCommitteeSpecTest{
+		Name:  "sequenced decided duties",
+		Tests: []*committee.CommitteeSpecTest{},
+	}
+
+	for _, numSequencedDuties := range []int{1, 2, 4} {
+		// TODO add 500
```
[the remark this answers is not in the document]
Comment (MatheusFranco99) [reply ssv/spectest/tests/committee/multipleduty/sequenced_decided_duties.go:21]: Yes

File: ssv/spectest/tests/committee/multipleduty/sequenced_happy_flow_duties.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		// TODO add 500
+		for _, numValidators := range []int{1, 30} {
+
+			ksMap := testingutils.KeySetMapForValidators(numValidators)
+			shareMap := testingutils.ShareMapFromKeySetMap(ksMap)
+
+			multiSpecTest.Tests = append(multiSpecTest.Tests, []*committee.CommitteeSpecTest{
+				{
+					Name:                   fmt.Sprintf("%v duties %v attestation", numSequencedDuties, numValidators),
+					Committee:              testingutils.BaseCommitteeWithRunnerSample(ksMap, testingutils.CommitteeRunnerWithShareMap(shareMap).(*ssv.CommitteeRunner)),
```
[the remark this answers is not in the document]
Comment (MatheusFranco99) [reply ssv/spectest/tests/committee/multipleduty/sequenced_happy_flow_duties.go:30]: There are ValidatorIndex -> KeySet and ValidatorIndex -> Share maps

File: ssv/spectest/tests/committee/multipleduty/failed_than_successful_duties.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import (
+	"fmt"
+
+	"github.com/ssvlabs/ssv-spec/ssv"
+	"github.com/ssvlabs/ssv-spec/ssv/spectest/tests"
+	"github.com/ssvlabs/ssv-spec/ssv/spectest/tests/committee"
+	"github.com/ssvlabs/ssv-spec/types/testingutils"
+)
+
+// FailedThanSuccessfulDuties decides a sequence of duties
```
[the remark this answers is not in the document]
Comment (MatheusFranco99) [reply ssv/spectest/tests/committee/multipleduty/failed_than_successful_duties.go:12]: This got here from a copy past 😅 
Will fix

File: ssv/spectest/tests/committee/multipleduty/sequenced_decided_duties.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"fmt"
+
+	"github.com/ssvlabs/ssv-spec/ssv"
+	"github.com/ssvlabs/ssv-spec/ssv/spectest/tests"
+	"github.com/ssvlabs/ssv-spec/ssv/spectest/tests/committee"
+	"github.com/ssvlabs/ssv-spec/types/testingutils"
+)
+
+// SequencedDecidedDuties decides a sequence of duties
```
[the remark this answers is not in the document]
Comment (MatheusFranco99) [reply ssv/spectest/tests/committee/multipleduty/sequenced_decided_duties.go:12]: Decided, in the sense of QBFT

File: types/testingutils/verifier.go [hunk 6]
```diff
@@ -75,6 +89,8 @@ func (v *testingVerifier) VerifySignatureForSigner(root [32]byte, signature []by
 }
 
 func (v *testingVerifier) HasSignature(operatorID types.OperatorID, root [32]byte, signature []byte) bool {
+	testingVerifierInstanceLock.Lock()
+	defer testingVerifierInstanceLock.Unlock()
```
[the remark this answers is not in the document]
Comment (MatheusFranco99) [reply types/testingutils/verifier.go:93]: yup

File: ssv/spectest/tests/committee/test.go (hunk 1 above)
Comment (MatheusFranco99) [reply ssv/spectest/tests/committee/test.go:57]: Yes

File: types/testingutils/verifier.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (GalRogozinski) [reply types/testingutils/verifier.go:93]: I meant to ask why 😅

File: ssv/spectest/tests/committee/multipleduty/sequenced_decided_duties.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (GalRogozinski) [reply ssv/spectest/tests/committee/multipleduty/sequenced_decided_duties.go:21]: I think you can finsih this in general and don't push the tests.json just yet...
we will try to solve this soon!

[the remark this answers is not in the document]
Comment (MatheusFranco99) [reply ssv/spectest/tests/committee/multipleduty/sequenced_decided_duties.go:21]: Ok! Will do that

File: types/testingutils/verifier.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (MatheusFranco99) [reply types/testingutils/verifier.go:93]: I've added a cache to the object that verifies the signatures and implemented the singleton pattern. But I forgot to add a lock on some methods.

**Pull request merged**